### PR TITLE
Feature/488 bpi subjects crud

### LINKED ISSUE #483: SRI Storage Component - SQL Calls to DB
## Overview

Would like to create SQL calls in Typescript that would allow the monolith to access and interact with the Postgres DB. 

### Reference

- [Database prep issue](https://github.com/eea-oasis/baseline/issues/482)
- [SRI Storage](https://github.com/eea-oasis/baseline/issues/478)

### Assumptions

- This issue assumes DB prep issue is completed. 

### Acceptance

- Enable the repo to make SQL calls to the Postgres DB
- Test this technology to ensure accuracy

## Tasks

- [x] Setup Node.js and node-postgres in the repo
- [x] Implement necessary functions to allow other components to make calls to the DB
- [x] Send queries to DB running in a docker container
- [ ] Create tests to make sure calls are returning and sending the right information
- [ ] Implement an ORM

### PR DESCRIPTION
**Pull request by ognjenkurtic**
# Description

Based on the agreement with @Ybittan, this one closes #483. I implemented Prisma as the ORM with a simple CREATE and GET operations for BpiSubjects, to set an example implementation other devs can continue building on. I ll continue working on BpiSubjects CRUD #488 on another branch. Added a bunch of TODOs as reminders for future issues.

Note 1: Didnt want to pollute the PR with formatting changes. Lets run formatter project-wide once no PRs open.

## Related Issue

#483
#488 

## Motivation and Context

Introduces an ORM abstracted from the business layer of the Bri. 

## How Has This Been Tested

npm install
npm run build
npm run start

Trigger post endpoint on http://localhost:3000/subjects
Trigger get endpoint http://localhost:3000/subjects/<newly_created_id>

Unit tests will follow as soon as TODO for mocking the repository is resolved.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Request to be added as a Code Owner/Maintainer

## Checklist

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ X] I commit to abide by the Responsibilities of Code Owners/Maintainers.

### CODE REVIEW

File: examples/bri-3/prisma/schema.prisma [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+generator client {
+  provider = "prisma-client-js"
+}
+
+datasource db {
+  provider = "sqlite"
+  url      = env("DATABASE_URL")
+}
+
+model BpiSubject {
```
Comment (Manik-Jain) [inline examples/bri-3/prisma/schema.prisma:13]: Just to understand this better, we have the similar declaration under migrations also. Does this model not define the table by default?
Also, are we going to do this pattern, I mean one migration file declaration and one in the model here for other models?

File: examples/bri-3/src/bri/identity/bpiSubjects/api/subjects.controller.ts [hunk 2]
```diff
@@ -23,4 +25,15 @@ export class SubjectController {
         )
       );
   }
+
+  @Put()
+  async UpdateBpiSubject(@Body() requestDto: UpdateBpiSubjectDto): Promise<BpiSubject> {
```
Comment (Manik-Jain) [inline examples/bri-3/src/bri/identity/bpiSubjects/api/subjects.controller.ts:30]: `updateBpiSubject`

File: examples/bri-3/src/bri/identity/bpiSubjects/capabilities/createBpiSubject/createBpiSubject.command.ts [hunk 3]
```diff
@@ -1,34 +1,7 @@
...
-  set _description(description: string) {
-    this.description = description;
-  }
-
-  set _publicKey(publicKey: string) {
-    this.publicKey = publicKey;
-  }
-
-  get() { return this}
+    public readonly name: string,
```
Comment (Manik-Jain) [inline examples/bri-3/src/bri/identity/bpiSubjects/capabilities/createBpiSubject/createBpiSubject.command.ts:3]: why are we changing these to public? We agreed in the previous PR that variables will be private.

File: examples/bri-3/src/bri/identity/bpiSubjects/capabilities/getBpiSubjectById/getBpiSubjectById.query.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export class GetBpiSubjectByIdQuery {
+  constructor(
+    public readonly id: string,
```
Comment (Manik-Jain) [inline examples/bri-3/src/bri/identity/bpiSubjects/capabilities/getBpiSubjectById/getBpiSubjectById.query.ts:3]: Question : why are we introducing public readonly variables? We agreed in the previous PR that variables will be private.

File: examples/bri-3/src/bri/identity/bpiSubjects/capabilities/updateBpiSubject/updateBpiSubject.command.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+export class UpdateBpiSubjectCommand {
```
Comment (Manik-Jain) [inline examples/bri-3/src/bri/identity/bpiSubjects/capabilities/updateBpiSubject/updateBpiSubject.command.ts:1]: Question : why are we introducing public readonly variables? We agreed in the previous PR that variables will be private.

**Review by Manik-Jain — CHANGES_REQUESTED** [review]
Added minor comments and questions. Looks good overall, with some TODO tasks.

File: examples/bri-3/prisma/schema.prisma (hunk 1 above)
Comment (ognjenkurtic) [reply examples/bri-3/prisma/schema.prisma:13]: This is just how prisma works. We only define (and change) our persistence models in schema.prisma, and then generate a migration and a prisma client out if it by using Prisma commands. I ll add a readme with the process in subsequent PRs. 

What is good about it is that the persistence model is completely separated from our domain objects, and we just map in both directions on the level of repositories.

File: examples/bri-3/src/bri/identity/bpiSubjects/capabilities/createBpiSubject/createBpiSubject.command.ts (hunk 3 above)
Comment (skosito) [reply examples/bri-3/src/bri/identity/bpiSubjects/capabilities/createBpiSubject/createBpiSubject.command.ts:3]: i believe this is same as just getter, without private setter, readonly making it so fields are only initialized in ctor

File: examples/bri-3/src/bri/identity/bpiSubjects/capabilities/getBpiSubjectById/getBpiSubjectByIdQuery.handler.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+@QueryHandler(GetBpiSubjectByIdQuery)
+export class GetBpiSubjectByIdQueryHandler implements IQueryHandler<GetBpiSubjectByIdQuery> {
+  constructor(
+    private readonly repo: BpiSubjectRepository,
+  ) {}
+
+  async execute(query: GetBpiSubjectByIdQuery) {
+    const bpiSubject = await this.repo.getBpiSubjectById(query.id);
+    return { // TODO:  // TODO: Write generic mapper domainObject -> DTO
```
Comment (skosito) [inline examples/bri-3/src/bri/identity/bpiSubjects/capabilities/getBpiSubjectById/getBpiSubjectByIdQuery.handler.ts:14]: lets make issue for generic mapper?

File: examples/bri-3/src/bri/identity/bpiSubjects/models/bpiSubject.ts [hunk 7]
```diff
@@ -1,24 +1,59 @@
...
+    this.name = name;
+    this.description = description;
+    this.type = type;
+    this.publicKey = publicKey;
+  }
+
+  public get id(): string {
+    return this._id;
+  }
+  public set id(value: string) {
```
Comment (skosito) [inline examples/bri-3/src/bri/identity/bpiSubjects/models/bpiSubject.ts:28]: why public setters?

File: examples/bri-3/src/bri/identity/bpiSubjects/persistence/bpiSubjects.repository.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import { BpiSubject } from '../models/bpiSubject';
+
+// Repositories are the only places that talk the Prisma language of models.
+// They are always mapped to and from domain objects so that the business layer of the application
+// does not have to care about the ORM.
+@Injectable()
+export class BpiSubjectRepository extends PrismaService {
+    async getBpiSubjectById(id: string): Promise<BpiSubject> {
+        const bpiSubjectModel = await this.bpiSubject.findUnique({ where: { id: id } });
+        return new BpiSubject( // TODO: Write generic mapper prismaModel -> domainObject
```
Comment (skosito) [inline examples/bri-3/src/bri/identity/bpiSubjects/persistence/bpiSubjects.repository.ts:12]: i could be wrong with this, but will a lot of mappers introduce a lot of boilerplate code
prisma model -> domain object -> dto object, and in most cases they are just same objects

**Review by skosito — DISMISSED** [review]
left some comments, but approved

File: examples/bri-3/prisma/migrations/migration_lock.toml [hunk 9]
```diff
@@ -0,0 +1,3 @@
+# Please do not edit this file manually
+# It should be added in your version-control system (i.e. Git)
+provider = "sqlite"
```
[the remark this answers is not in the document]
Comment (ognjenkurtic) [reply examples/bri-3/prisma/migrations/migration_lock.toml:3]: yes, as soon as #482 is done.

File: examples/bri-3/prisma/prisma.service.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import { Injectable, OnModuleInit, INestApplication } from '@nestjs/common'
+import { PrismaClient } from '@prisma/client'
+
+// TODO: Not sure if this is a good location for the prisma service.
+// Decided to put it here because it is Prisma ORM specific, so BRI should not be aware of it.
```
[the remark this answers is not in the document]
Comment (ognjenkurtic) [reply examples/bri-3/prisma/prisma.service.ts]: Ok, will remove. I was refering to bri business logic located under src when i wrote 'bri'

File: examples/bri-3/src/bri/identity/bpiSubjects/capabilities/getBpiSubjectById/getBpiSubjectByIdQuery.handler.ts (hunk 6 above)
Comment (ognjenkurtic) [reply examples/bri-3/src/bri/identity/bpiSubjects/capabilities/getBpiSubjectById/getBpiSubjectByIdQuery.handler.ts:14]: Whoever has time first.

File: examples/bri-3/src/bri/identity/bpiSubjects/persistence/bpiSubjects.repository.ts (hunk 8 above)
Comment (ognjenkurtic) [reply examples/bri-3/src/bri/identity/bpiSubjects/persistence/bpiSubjects.repository.ts:12]: That is why is must be a generic mapper. Usually something along the line of: Object.assign(this). This way it is just a single line of code that does the mapping. Domain objects will not be the same, they ll contain methods for the execution of  the business logic. Persistence model might contain fields such as CreatedAt,  ModifiedAt that will not be mapped to the domain object as they are there only for audit purposes.

File: examples/bri-3/src/bri/identity/bpiSubjects/models/bpiSubject.ts (hunk 7 above)
Comment (ognjenkurtic) [reply examples/bri-3/src/bri/identity/bpiSubjects/models/bpiSubject.ts:28]: good point, added them on autopilot (not github autopilot :)). Removing..

File: examples/bri-3/package.json [hunk 11]
```diff
@@ -45,6 +46,7 @@
     "eslint-plugin-prettier": "^4.0.0",
     "jest": "28.1.2",
     "prettier": "^2.3.2",
+    "prisma": "^4.3.1",
```
[the remark this answers is not in the document]
Comment (skosito) [reply examples/bri-3/package.json:49]: there are 2 prisma packages, prisma installed as dev dependency and @prisma/client installed as regular dependency, why do we need both?
i also see @prisma/migrate exists, do we need that too?

[the remark this answers is not in the document]
Comment (ognjenkurtic) [reply examples/bri-3/package.json:49]: Prisma is the CLI package used to run the commands. Prisma client holds the generated models and is the 'service' we inject into the app. Not sure what prisma/migrate does, didn't have a need for it yet.

File: examples/bri-3/src/bri/identity/bpiSubjects/capabilities/getBpiSubjectById/getBpiSubjectByIdQuery.handler.ts (hunk 6 above)
Comment (Kasshern) [reply examples/bri-3/src/bri/identity/bpiSubjects/capabilities/getBpiSubjectById/getBpiSubjectByIdQuery.handler.ts:14]: This task is very important and therefore requires a double TODO.

**Review by Kasshern — APPROVED** [review]

**Review by Manik-Jain — APPROVED** [review]

**Pull request merged**